File: scripts/ferretd.mjs
import { createHash, randomUUID } from 'node:crypto';
import {
  createReadStream,
  createWriteStream,
} from 'node:fs';
import {
  chmod,
  copyFile,
  lstat,
  mkdir,
  readFile,
  rename,
  rm,
  writeFile,
} from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { Readable } from 'node:stream';
import { pipeline } from 'node:stream/promises';
import { createGunzip } from 'node:zlib';

import tarStream from 'tar-stream';
import yauzl from 'yauzl';

const releaseRepository = 'MontFerret/ferretd';
const checksumAsset = 'ferretd_checksums.txt';
const maximumBinarySize = 128 * 1024 * 1024;
const versionPattern = /^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?$/u;

export async function readFerretdVersion(repositoryRoot) {
  const manifestPath = join(repositoryRoot, 'ferretd.json');
  let manifest;

  try {
    manifest = JSON.parse(await readFile(manifestPath, 'utf8'));
  } catch (error) {
    throw new Error(`Cannot read ${manifestPath}: ${formatError(error)}`);
  }

  if (
    manifest === null ||
    typeof manifest !== 'object' ||
    Array.isArray(manifest) ||
    Object.keys(manifest).length !== 1 ||
    typeof manifest.ferretd !== 'string' ||
    !versionPattern.test(manifest.ferretd)
  ) {
    throw new Error(
      `${manifestPath} must contain exactly one valid "ferretd" version`,
    );
  }

  return manifest.ferretd;
}

export function parseChecksums(contents) {
  const checksums = new Map();

  for (const [index, rawLine] of contents.split(/\r?\n/u).entries()) {
    if (rawLine === '') {
      continue;
    }

    const match = /^([0-9a-f]{64})  ([^/\\]+)$/u.exec(rawLine);
    if (match === null) {
      throw new Error(`Invalid checksum line ${index + 1}`);
    }

    const [, digest, name] = match;
    if (checksums.has(name)) {
      throw new Error(`Duplicate checksum entry for ${name}`);
    }

    checksums.set(name, digest);
  }

  if (checksums.size === 0) {
    throw new Error('Checksum manifest is empty');
  }

  return checksums;
}

export async function sha256File(path) {
  const hash = createHash('sha256');
  await pipeline(createReadStream(path), hash);

  return hash.digest('hex');
}

export function releaseAssetUrl(version, assetName) {
  validateFerretdVersion(version);
  if (!/^[0-9A-Za-z._-]+$/u.test(assetName)) {
    throw new Error(`Invalid ferretd release asset: ${assetName}`);
  }

  return (
    `https://github.com/${releaseRepository}/releases/download/` +
    `v${version}/${assetName}`
  );
}

export function sourceArchiveUrl(version) {
  validateFerretdVersion(version);

  return (
    `https://github.com/${releaseRepository}/archive/refs/tags/` +
    `v${version}.tar.gz`
  );
}

export async function acquireFerretd({
  repositoryRoot,
  target,
  artifact,
  archiveType,
  binaryName,
  unix,
  fetchImplementation = globalThis.fetch,
}) {
  validateAcquisitionInput({
    target,
    artifact,
    archiveType,
    binaryName,
  });

  if (typeof fetchImplementation !== 'function') {
    throw new Error('A fetch implementation is required');
  }

  const version = await readFerretdVersion(repositoryRoot);
  const cacheRoot = join(
    repositoryRoot,
    '.dist',
    'ferretd',
    version,
    target,
  );
  await mkdir(cacheRoot, { recursive: true });

  const checksumPath = join(cacheRoot, checksumAsset);
  if (!(await isFile(checksumPath))) {
    await downloadFile(
      releaseAssetUrl(version, checksumAsset),
      checksumPath,
      fetchImplementation,
    );
  }

  const checksums = parseChecksums(await readFile(checksumPath, 'utf8'));
  const expectedDigest = checksums.get(artifact);
  if (expectedDigest === undefined) {
    throw new Error(
      `${checksumAsset} does not contain ${artifact} for ${target}`,
    );
  }

  const archivePath = join(cacheRoot, artifact);
  if (!(await isFile(archivePath))) {
    await downloadFile(
      releaseAssetUrl(version, artifact),
      archivePath,
      fetchImplementation,
    );
  }

  const actualDigest = await sha256File(archivePath);
  if (actualDigest !== expectedDigest) {
    await rm(archivePath, { force: true });

    throw new Error(
      `Checksum mismatch for ${artifact}: expected ${expectedDigest}, got ${actualDigest}`,
    );
  }

  const binary = await extractBinary(
    archivePath,
    archiveType,
    binaryName,
  );
  const extractedPath = join(cacheRoot, 'extracted', binaryName);
  await writeFileAtomic(extractedPath, binary, unix ? 0o755 : 0o644);

  return {
    archivePath,
    binaryPath: extractedPath,
    digest: actualDigest,
    version,
  };
}

export async function copyFileAtomic(source, destination, mode) {
  await mkdir(dirname(destination), { recursive: true });
  const temporary = `${destination}.partial-${process.pid}-${randomUUID()}`;

  try {
    await copyFile(source, temporary);
    await chmod(temporary, mode);
    await rm(destination, { force: true });
    await rename(temporary, destination);
  } catch (error) {
    await rm(temporary, { force: true });
    throw error;
  }
}

async function downloadFile(url, destination, fetchImplementation) {
  await mkdir(dirname(destination), { recursive: true });
  const temporary = `${destination}.partial-${process.pid}-${randomUUID()}`;

  try {
    const response = await fetchImplementation(url, { redirect: 'follow' });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status} ${response.statusText}`);
    }

    if (response.body === null) {
      throw new Error('response has no body');
    }

    await pipeline(
      Readable.fromWeb(response.body),
      createWriteStream(temporary, { flags: 'wx', mode: 0o600 }),
    );
    await rename(temporary, destination);
  } catch (error) {
    await rm(temporary, { force: true });
    throw new Error(`Cannot download ${url}: ${formatError(error)}`);
  }
}

async function extractBinary(archivePath, archiveType, binaryName) {
  if (archiveType === 'tar.gz') {
    return extractTarGzipBinary(archivePath, binaryName);
  }

  if (archiveType === 'zip') {
    return extractZipBinary(archivePath, binaryName);
  }

  throw new Error(`Unsupported archive type: ${archiveType}`);
}

async function extractTarGzipBinary(archivePath, binaryName) {
  const extract = tarStream.extract();
  let binary;

  extract.on('entry', (header, stream, next) => {
    if (header.name !== binaryName) {
      stream.on('error', (error) => extract.destroy(error));
      stream.on('end', next);
      stream.resume();
      return;
    }

    if (binary !== undefined) {
      stream.resume();
      extract.destroy(new Error(`Archive contains duplicate ${binaryName}`));
      return;
    }
    if (header.type !== 'file') {
      stream.resume();
      extract.destroy(new Error(`${binaryName} is not a regular file`));
      return;
    }

    collectStream(stream, maximumBinarySize).then(
      (contents) => {
        binary = contents;
        next();
      },
      (error) => extract.destroy(error),
    );
  });

  try {
    await pipeline(
      createReadStream(archivePath),
      createGunzip(),
      extract,
    );
  } catch (error) {
    throw new Error(
      `Cannot extract ${archivePath}: ${formatError(error)}`,
    );
  }

  if (binary === undefined) {
    throw new Error(`${archivePath} does not contain ${binaryName}`);
  }

  return binary;
}

async function extractZipBinary(archivePath, binaryName) {
  const zip = await openZip(archivePath);
  let binary;

  try {
    while (true) {
      const entry = await readZipEntry(zip);
      if (entry === undefined) {
        break;
      }
      if (entry.fileName !== binaryName) {
        continue;
      }
      if (binary !== undefined) {
        throw new Error(`Archive contains duplicate ${binaryName}`);
      }
      if (/\/$/u.test(entry.fileName)) {
        throw new Error(`${binaryName} is not a regular file`);
      }
      const mode = (entry.externalFileAttributes >>> 16) & 0xffff;
      const fileType = mode & 0o170000;
      if (fileType !== 0 && fileType !== 0o100000) {
        throw new Error(`${binaryName} is not a regular file`);
      }
      if (entry.uncompressedSize > maximumBinarySize) {
        throw new Error(`${binaryName} exceeds the extraction size limit`);
      }

      const stream = await openZipEntry(zip, entry);
      binary = await collectStream(stream, maximumBinarySize);
    }
  } catch (error) {
    zip.close();
    throw new Error(
      `Cannot extract ${archivePath}: ${formatError(error)}`,
    );
  }

  zip.close();
  if (binary === undefined) {
    throw new Error(`${archivePath} does not contain ${binaryName}`);
  }

  return binary;
}

function openZip(path) {
  return new Promise((resolve, reject) => {
    yauzl.open(path, { lazyEntries: true }, (error, zip) => {
      if (error !== null) {
        reject(error);
        return;
      }

      resolve(zip);
    });
  });
}

function readZipEntry(zip) {
  return new Promise((resolve, reject) => {
    const onEntry = (entry) => {
      cleanup();
      resolve(entry);
    };
    const onEnd = () => {
      cleanup();
      resolve(undefined);
    };
    const onError = (error) => {
      cleanup();
      reject(error);
    };
    const cleanup = () => {
      zip.off('entry', onEntry);
      zip.off('end', onEnd);
      zip.off('error', onError);
    };

    zip.once('entry', onEntry);
    zip.once('end', onEnd);
    zip.once('error', onError);
    zip.readEntry();
  });
}

function openZipEntry(zip, entry) {
  return new Promise((resolve, reject) => {
    zip.openReadStream(entry, (error, stream) => {
      if (error !== null) {
        reject(error);
        return;
      }

      resolve(stream);
    });
  });
}

async function collectStream(stream, maximumSize) {
  const chunks = [];
  let size = 0;

  for await (const chunk of stream) {
    size += chunk.length;
    if (size > maximumSize) {
      throw new Error('Archive entry exceeds the extraction size limit');
    }

    chunks.push(chunk);
  }

  return Buffer.concat(chunks, size);
}

async function writeFileAtomic(destination, contents, mode) {
  await mkdir(dirname(destination), { recursive: true });
  const temporary = `${destination}.partial-${process.pid}-${randomUUID()}`;

  try {
    await writeFile(temporary, contents, { flag: 'wx', mode });
    await chmod(temporary, mode);
    await rm(destination, { force: true });
    await rename(temporary, destination);
  } catch (error) {
    await rm(temporary, { force: true });
    throw error;
  }
}

async function isFile(path) {
  try {
    return (await lstat(path)).isFile();
  } catch (error) {
    if (error?.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}

function validateAcquisitionInput({
  target,
  artifact,
  archiveType,
  binaryName,
}) {
  if (!/^[a-z0-9-]+$/u.test(target)) {
    throw new Error(`Invalid distribution target: ${target}`);
  }

  if (!/^[0-9A-Za-z._-]+$/u.test(artifact)) {
    throw new Error(`Invalid release artifact: ${artifact}`);
  }

  if (archiveType !== 'tar.gz' && archiveType !== 'zip') {
    throw new Error(`Unsupported archive type: ${archiveType}`);
  }

  if (binaryName !== 'ferretd' && binaryName !== 'ferretd.exe') {
    throw new Error(`Invalid daemon executable name: ${binaryName}`);
  }
}

function validateFerretdVersion(version) {
  if (!versionPattern.test(version)) {
    throw new Error(`Invalid ferretd version: ${version}`);
  }
}

function formatError(error) {
  return error instanceof Error ? error.message : String(error);
}

File: tools/editorium/ferretd.go
package main

import (
	"archive/tar"
	"archive/zip"
	"bufio"
	"bytes"
	"compress/gzip"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"golang.org/x/mod/semver"
)

const (
	ferretdRepository = "MontFerret/ferretd"
	checksumAsset     = "ferretd_checksums.txt"
	maximumBinarySize = 128 * 1024 * 1024
)

var checksumLinePattern = regexp.MustCompile(`^([0-9a-f]{64})  ([^/\\]+)$`)

type (
	httpDoer interface {
		Do(*http.Request) (*http.Response, error)
	}

	acquiredFerretd struct {
		ArchivePath string
		BinaryPath  string
		Digest      string
		Version     string
	}
)

func readFerretdVersion(root string) (string, error) {
	path := filepath.Join(root, "ferretd.json")
	contents, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("cannot read %s: %w", path, err)
	}

	decoder := json.NewDecoder(bytes.NewReader(contents))
	opening, err := decoder.Token()
	if err != nil {
		return "", fmt.Errorf("cannot read %s: %w", path, err)
	}

	if opening != json.Delim('{') {
		return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
	}

	var value json.RawMessage
	keys := 0
	seen := make(map[string]struct{})
	for decoder.More() {
		keyToken, tokenErr := decoder.Token()
		if tokenErr != nil {
			return "", fmt.Errorf("cannot read %s: %w", path, tokenErr)
		}

		key, ok := keyToken.(string)
		if !ok {
			return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
		}

		if _, duplicate := seen[key]; duplicate {
			return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
		}

		seen[key] = struct{}{}
		keys++

		var raw json.RawMessage
		if err := decoder.Decode(&raw); err != nil {
			return "", fmt.Errorf("cannot read %s: %w", path, err)
		}

		if key == "ferretd" {
			value = raw
		}
	}

	closing, err := decoder.Token()
	if err != nil {
		return "", fmt.Errorf("cannot read %s: %w", path, err)
	}

	if closing != json.Delim('}') {
		return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
	}

	if trailingErr := decoder.Decode(new(any)); trailingErr != io.EOF {
		return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
	}

	if keys != 1 || value == nil {
		return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
	}

	var version string
	if err := json.Unmarshal(value, &version); err != nil || !validVersion(version) {
		return "", fmt.Errorf("%s must contain exactly one valid \"ferretd\" version", path)
	}

	return version, nil
}

func validVersion(version string) bool {
	return version != "" && semver.IsValid("v"+version)
}

func releaseAssetURL(version, asset string) (string, error) {
	if !validVersion(version) {
		return "", fmt.Errorf("invalid ferretd version: %s", version)
	}

	if matched, _ := regexp.MatchString(`^[0-9A-Za-z._-]+$`, asset); !matched {
		return "", fmt.Errorf("invalid ferretd release asset: %s", asset)
	}

	return fmt.Sprintf("https://github.com/%s/releases/download/v%s/%s", ferretdRepository, version, asset), nil
}

func sourceArchiveURL(version string) (string, error) {
	if !validVersion(version) {
		return "", fmt.Errorf("invalid ferretd version: %s", version)
	}

	return fmt.Sprintf("https://github.com/%s/archive/refs/tags/v%s.tar.gz", ferretdRepository, version), nil
}

func parseChecksums(contents string) (map[string]string, error) {
	result := make(map[string]string)
	scanner := bufio.NewScanner(strings.NewReader(contents))
	line := 0

	for scanner.Scan() {
		line++
		raw := strings.TrimSuffix(scanner.Text(), "\r")
		if raw == "" {
			continue
		}

		matches := checksumLinePattern.FindStringSubmatch(raw)
		if matches == nil {
			return nil, fmt.Errorf("invalid checksum line %d", line)
		}

		if _, exists := result[matches[2]]; exists {
			return nil, fmt.Errorf("duplicate checksum entry for %s", matches[2])
		}

		result[matches[2]] = matches[1]
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	if len(result) == 0 {
		return nil, fmt.Errorf("checksum manifest is empty")
	}

	return result, nil
}

func sha256File(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}

	defer file.Close()

	hash := sha256.New()

	if _, err := io.Copy(hash, file); err != nil {
		return "", err
	}

	return hex.EncodeToString(hash.Sum(nil)), nil
}

func acquireFerretd(ctx context.Context, root string, target ferretdTarget, client httpDoer) (acquiredFerretd, error) {
	if client == nil {
		client = http.DefaultClient
	}

	version, err := readFerretdVersion(root)
	if err != nil {
		return acquiredFerretd{}, err
	}

	cacheRoot := filepath.Join(root, ".dist", "ferretd", version, target.ID)
	if err := os.MkdirAll(cacheRoot, 0o755); err != nil {
		return acquiredFerretd{}, err
	}

	checksumPath := filepath.Join(cacheRoot, checksumAsset)
	if !regularFile(checksumPath) {
		url, urlErr := releaseAssetURL(version, checksumAsset)
		if urlErr != nil {
			return acquiredFerretd{}, urlErr
		}

		if err := downloadFile(ctx, client, url, checksumPath); err != nil {
			return acquiredFerretd{}, err
		}
	}

	checksumContents, err := os.ReadFile(checksumPath)
	if err != nil {
		return acquiredFerretd{}, err
	}

	checksums, err := parseChecksums(string(checksumContents))
	if err != nil {
		return acquiredFerretd{}, err
	}

	expected, ok := checksums[target.Artifact]
	if !ok {
		return acquiredFerretd{}, fmt.Errorf("%s does not contain %s for %s", checksumAsset, target.Artifact, target.ID)
	}

	archivePath := filepath.Join(cacheRoot, target.Artifact)
	if !regularFile(archivePath) {
		url, urlErr := releaseAssetURL(version, target.Artifact)
		if urlErr != nil {
			return acquiredFerretd{}, urlErr
		}

		if err := downloadFile(ctx, client, url, archivePath); err != nil {
			return acquiredFerretd{}, err
		}
	}

	actual, err := sha256File(archivePath)
	if err != nil {
		return acquiredFerretd{}, err
	}

	if actual != expected {
		_ = os.Remove(archivePath)

		return acquiredFerretd{}, fmt.Errorf("checksum mismatch for %s: expected %s, got %s", target.Artifact, expected, actual)
	}

	contents, err := extractBinary(archivePath, target)
	if err != nil {
		return acquiredFerretd{}, err
	}

	mode := os.FileMode(0o644)
	if target.Unix {
		mode = 0o755
	}

	binaryPath := filepath.Join(cacheRoot, "extracted", target.BinaryName)
	if err := writeFileAtomic(binaryPath, contents, mode); err != nil {
		return acquiredFerretd{}, err
	}

	return acquiredFerretd{ArchivePath: archivePath, BinaryPath: binaryPath, Digest: actual, Version: version}, nil
}

func downloadFile(ctx context.Context, client httpDoer, url, destination string) error {
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}

	response, err := client.Do(request)
	if err != nil {
		return fmt.Errorf("cannot download %s: %w", url, err)
	}

	defer response.Body.Close()

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("cannot download %s: HTTP %s", url, response.Status)
	}

	if err := os.MkdirAll(filepath.Dir(destination), 0o755); err != nil {
		return err
	}

	temporary, err := os.CreateTemp(filepath.Dir(destination), filepath.Base(destination)+".partial-")
	if err != nil {
		return err
	}

	temporaryPath := temporary.Name()
	defer os.Remove(temporaryPath)

	if _, err := io.Copy(temporary, response.Body); err != nil {
		temporary.Close()
		return fmt.Errorf("cannot download %s: %w", url, err)
	}

	if err := temporary.Close(); err != nil {
		return err
	}

	if err := replaceFile(temporaryPath, destination); err != nil {
		return err
	}

	return nil
}

func extractBinary(path string, target ferretdTarget) ([]byte, error) {
	switch target.ArchiveType {
	case "tar.gz":
		return extractTarBinary(path, target.BinaryName)
	case "zip":
		return extractZipBinary(path, target.BinaryName)
	default:
		return nil, fmt.Errorf("unsupported archive type: %s", target.ArchiveType)
	}
}

func extractTarBinary(path, binaryName string) ([]byte, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}

	defer file.Close()

	gzipReader, err := gzip.NewReader(file)
	if err != nil {
		return nil, fmt.Errorf("cannot extract %s: %w", path, err)
	}

	defer gzipReader.Close()

	reader := tar.NewReader(gzipReader)
	var binary []byte
	for {
		header, nextErr := reader.Next()
		if nextErr == io.EOF {
			break
		}

		if nextErr != nil {
			return nil, fmt.Errorf("cannot extract %s: %w", path, nextErr)
		}

		if header.Name != binaryName {
			continue
		}

		if binary != nil {
			return nil, fmt.Errorf("archive contains duplicate %s", binaryName)
		}

		if header.Typeflag != tar.TypeReg {
			return nil, fmt.Errorf("%s is not a regular file", binaryName)
		}

		var buffer bytes.Buffer
		if err := copyLimited(&buffer, reader, maximumBinarySize, binaryName); err != nil {
			return nil, err
		}

		binary = buffer.Bytes()
	}

	if binary == nil {
		return nil, fmt.Errorf("%s does not contain %s", path, binaryName)
	}

	return binary, nil
}

func extractZipBinary(path, binaryName string) ([]byte, error) {
	archive, err := zip.OpenReader(path)
	if err != nil {
		return nil, fmt.Errorf("cannot extract %s: %w", path, err)
	}

	defer archive.Close()

	var binary []byte
	for _, file := range archive.File {
		if file.Name != binaryName {
			continue
		}

		if binary != nil {
			return nil, fmt.Errorf("archive contains duplicate %s", binaryName)
		}

		if file.FileInfo().IsDir() || !file.Mode().IsRegular() || file.UncompressedSize64 > maximumBinarySize {
			return nil, fmt.Errorf("%s is not a bounded regular file", binaryName)
		}

		reader, openErr := file.Open()
		if openErr != nil {
			return nil, openErr
		}

		var buffer bytes.Buffer
		copyErr := copyLimited(&buffer, reader, maximumBinarySize, binaryName)
		closeErr := reader.Close()

		if copyErr != nil {
			return nil, copyErr
		}

		if closeErr != nil {
			return nil, closeErr
		}

		binary = buffer.Bytes()
	}

	if binary == nil {
		return nil, fmt.Errorf("%s does not contain %s", path, binaryName)
	}

	return binary, nil
}

func writeFileAtomic(destination string, contents []byte, mode os.FileMode) error {
	if err := os.MkdirAll(filepath.Dir(destination), 0o755); err != nil {
		return err
	}

	temporary, err := os.CreateTemp(filepath.Dir(destination), filepath.Base(destination)+".partial-")
	if err != nil {
		return err
	}

	temporaryPath := temporary.Name()
	defer os.Remove(temporaryPath)

	if _, err := temporary.Write(contents); err != nil {
		temporary.Close()

		return err
	}

	if err := temporary.Chmod(mode); err != nil {
		temporary.Close()

		return err
	}

	if err := temporary.Close(); err != nil {
		return err
	}

	return replaceFile(temporaryPath, destination)
}

func copyFileAtomic(source, destination string, mode os.FileMode) error {
	contents, err := os.ReadFile(source)
	if err != nil {
		return err
	}

	return writeFileAtomic(destination, contents, mode)
}

func replaceFile(source, destination string) error {
	if err := os.Remove(destination); err != nil && !os.IsNotExist(err) {
		return err
	}

	return os.Rename(source, destination)
}

func regularFile(path string) bool {
	info, err := os.Lstat(path)

	return err == nil && info.Mode().IsRegular()
}
